### DISCUSSION TITLE
phenotypic mean & signing the estimated (genomic) breeding values by external program

### DISCUSSION THREAD

**Discussion by Hasan-Baneh** [gaynorr/AlphaSimR#82] [Q&A]
Dear developers,

Thanks for providing the package. 
I have  two questions: 
First, how to change the phenotypic scale of the traits? If the program samples the effects with the user defined means and variances, therefore, the phenotype mean for each trait should be summation the mean of all simulated effects. Is there any function to change the phenotype mean of the traits? or the users should do it by themselves, e.g., add intercept to the simulated phenotypic values of each trait?

And the second question is about assigning the estimated (genomic) breeding values by external program in a multi-trait simulation scheme. For example in my simulation [a multi[n=4] correlated traits], when I estimated gebv for that traits and saving in a matrix [nInd X No.traits] and assigned this matrix to simulated population using [pop_TMP@ebv = matrix(my_GEBVs)]. It did not work well and the dimension of pop_TMP@ebv was not right to that of the original matrix . When calling pop_TMP@ebv, it returns  a vector with the length of [nInd* No.Traits]. While it should be a matrix with dimension of [nrow=nInd X ncol=No.Traits].

Any ideas?

Best regards

**Comment by gaynorr** [comment]
Hi Hasan-Baneh,

The software scales QTL effects and adds an intercept to achieve a specified genetic mean and variance in the initial population. By default, the variance pertains to the additive genetic variance, but this can be changed to the total genetic variance by using the option useVarA=FALSE in the addTrait_ function. The expected phenotypic variance is then just the sum of this total genetic variance and the residual variance added to the trait via the use of `SimParam$setVarE` or `setPheno`. Both of these functions take either varE, h2, or H2 as an argument. The h2 and H2 arguments are just alternative ways to set varE that automatically calculate the right level of varE to achieve you target heritability. Thus, you are indirectly setting the phenotypic variance based on your decisions above. 

For the second issue, I suggest examining the dimensions of matrix(my_GEBVs). When manually assigning a value to this slot, R only checks that the object you are assigning is a matrix. It doesn't check whether it has the correct dimensions. I suspect the dimensions were wrong on the object being assigned. You can manually trigger a check on the matrix dimensions by running `validObject(pop_TMP)`. It should come back to you with this message:
"Error in validObject(pop_TMP) : invalid class “Pop” object: nInd!=nrow(ebv)"

Here is a script to show some of things I mentioned above. Hopefully, it will help if any of my comments aren't clear.
```
library(AlphaSimR)

# Create founder haplotypes
founderPop = runMacs(nInd=100, nChr=10, segSites=1500)

# Set simulation parameters for 4 pleiotropic traits
# Each trait will be controlled by 1,000 QTL per chromosome
# Mean will be 0 and genetic variance will be 1 for all traits
# All traits will have 0.5 genetic correlation with each other
# The heritability for each trait is set to 0.3
SP = SimParam$
  new(founderPop = founderPop)$
  addTraitA(nQtlPerChr = 1000,
            mean = rep(0, 4), # Means for 4 traits
            var = rep(1, 4), # Variances for 4 traits
            corA = 0.5*diag(4)+0.5)$  # Genetic correlations for 4 traits
  setVarE(h2 = rep(0.3, 4))$ # heritabities for 4 traits
  addSnpChip(nSnpPerChr = 500) # SNPs per chromosome for fitting GS model

# Create an initial population
pop = newPop(founderPop)

# Observed phenotypic variance
varP(pop)

# Generate a new phenotype to observe stochastic variability 
# repeat this several times
pop = setPheno(pop)
varP(pop)

# Expected variability is the genetic variance of the population
# Plus the residual variance
varG(pop) # Genetic variance of population
SP$varE # Residual variance assigned to achieve requested heritability

# Note that varE is a vector because we didn't model correlated error, 
# so you have to turn it into a matrix to get the expected phenotypic variance.
# Expected phenotypic variance
varG(pop) + diag(SP$varE)

# Fit a multivariate GS model manually using an AlphaSimR solver
# This requires constructing design matrices for the solver
# A similar model can be fit with the RRBLUP function without needing to 
# manually construct the matrices

# Phenotype matrix
Y = pheno(pop) 

# Fixed effect design matrix, intercept only
X = matrix(1, nrow=nInd(pop), ncol=1) 

# Random effect design matrix, fitting a GBLUP model
Z = diag(nInd(pop))

# G matrix for random effects
M = pullSnpGeno(pop)
p = colMeans(M)/2 # Allele frequency
G = tcrossprod(scale(M, scale=TRUE)) / (2*sum(p*(1-p))) 

# Fit a multivariate model
ans = solveMVM(Y=Y, X=X, Z=Z, K=G)

# Estimated variance components
# Not expected to be good due to the small size
ans$Vu
ans$Ve

# Assign BLUP values to the ebv slot
pop@ebv = ans$u
dim(pop@ebv) # nInd x 4 
validObject(pop) # TRUE

# Examine correlations between genetic values and EBVs
cor(gv(pop), ebv(pop))
```
